### PR TITLE
Generate function bodies with nondet return values

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/langapi/language.h [hunk 1]
```diff
@@ -70,8 +70,7 @@ class languaget:public messaget
   /// complete. Functions introduced here are visible to lazy loading and
   /// can influence its decisions (e.g. picking the types of input parameters
   /// and globals), whereas anything introduced during `final` cannot.
-  virtual bool generate_support_functions(
-    symbol_tablet &symbol_table)=0;
+  virtual bool generate_support_functions(symbol_tablet &symbol_table) = 0;
```
Comment (smowton) [inline src/langapi/language.h]: Unrelated restyle in commit

File: src/langapi/language_file.cpp [hunk 2]
```diff
@@ -10,6 +10,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <fstream>
 
+#include <util/object_factory_parameters.h>
```
Comment (smowton) [inline src/langapi/language_file.cpp]: Adding an include but nothing that uses it (in this commit)?

File: src/ansi-c/c_nondet_symbol_factory.cpp [hunk 3]
```diff
@@ -22,27 +22,6 @@ Author: Diffblue Ltd.
 
 #include <goto-programs/goto_functions.h>
 
-/// Create a new temporary static symbol
-/// \param symbol_table: The symbol table to create the symbol in
-/// \param loc: The location to assign to the symbol
-/// \param type: The type of symbol to create
-/// \param static_lifetime: Whether the symbol should have a static lifetime
-/// \param prefix: The prefix to use for the symbol's basename
```
Comment (smowton) [inline src/ansi-c/c_nondet_symbol_factory.cpp]: Why is this in this PR?

File: regression/cbmc/pointer-function-parameters-struct-simple-recursion-2/main.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+void func(st_t *p)
+{
+  assert(p != NULL);
+  assert(p != &dummy);
+
+  assert(p->prev != NULL);
+  assert(p->prev != &dummy);
+
+  assert(p->next != NULL);
+  assert(p->next != &dummy);
```
Comment (smowton) [inline regression/cbmc/pointer-function-parameters-struct-simple-recursion-2/main.c]: Also assert p->next != p->prev?

File: src/ansi-c/c_nondet_symbol_factory.cpp [hunk 5]
```diff
@@ -181,7 +224,6 @@ void symbol_factoryt::gen_nondet_init(
 /// \param base_name: The name to use for the symbol created
 /// \param type: The type for the symbol created
 /// \param loc: The location to assign to generated code
-/// \param allow_null: Whether to allow a null value when type is a pointer
```
Comment (smowton) [inline src/ansi-c/c_nondet_symbol_factory.cpp]: does removing this unused parameter belong in a different commit?

File: src/util/allocate_objects.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+/// \param assignments: The code block to add code to.
+/// \param target_expr: The expression which we are allocating a symbol for.
+/// \param allocate_type: Type of the object allocated
+/// \param alloc_type: Allocation type (global, local or dynamic)
+/// \param symbols_created: Created symbols to be declared by the caller
+/// \return An expression denoting the address of the newly allocated object.
+exprt allocate_objectst::allocate_object(
+  code_blockt &assignments,
+  const exprt &target_expr,
+  const typet &allocate_type,
```
Comment (smowton) [inline src/util/allocate_objects.cpp]: allocate_type and alloc_type are very similar names

File: src/util/allocate_objects.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+#include <util/type.h>
+
+/// Selects the kind of allocation used by the object factories
+enum class allocation_typet
+{
+  /// Allocate global objects
+  GLOBAL,
+  /// Allocate local stacked objects
+  LOCAL,
+  /// Allocate dynamic objects (using MALLOC)
```
Comment (smowton) [inline src/util/allocate_objects.h]: now using ALLOCATE

**Review by smowton — COMMENTED** [review]
Cursory review because this is *big*. For a more thorough look over I recommend introducing this in stages

**Comment by danpoe** [comment]
@smowton Thanks for the review! The PR is based on another open PR (#3455) and in this one only the last two commits are new (see PR description).

**Comment by kroening** [comment]
Should this live in goto-instrument? It's an instrumentation.

**Review by kroening — CHANGES_REQUESTED** [review]
Should be in goto-instrument!

**Comment by danpoe** [comment]
@kroening One thing this PR does is move `generate_function_bodies.h/cpp` from `goto-programs` to `goto-instrument`. However I left the C object factory in `ansi-c` (just like the Java object factory is in `java_bytecode`). Are you suggesting to move the C object factory to `goto-instrument` as well?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f2506f96a48257944df5ef908182c824648b30ee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93055935

**Comment by danpoe** [comment]
@kroening This is now done in goto-instrument.

**Review by chrisr-diffblue — APPROVED** [review]
LGTM

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cc4baa5b3b7d3c5354806d54fb76b343e5ca2a4e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98196114
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5fa9452d56904051f480d5595ffdf64432e80018).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98301475

**Pull request merged**